File: Solutions/0198-house-robber/solution.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int len=nums.size();
        if(len==0) return 0;
        int dp[len+2];
        memset(dp,0,sizeof(dp));
        dp[0]=nums[0];
        if(len>1) dp[1]=max(dp[0],nums[1]);
        for(int i=2;i<nums.size();i++)
        {
            dp[i]=max(dp[i-1],dp[i-2]+nums[i]);
        }
        
        return dp[len-1];
    }
};

File: Solutions/0198-house-robber/solution.go
func rob(nums []int) int {
    n := len(nums)
    dp := make([]int, n)
    
    dp[0] = nums[0]
    if n > 1 {
        dp[1] = max(nums[1] , nums[0])
    }
    if n > 2 {
        dp[2] = max( nums[2] + nums[0] , nums[1] )  
    }

    for i:=3;i<n;i++ {
        dp[i] = max(nums[i]+dp[i-2], dp[i-1])
    }

    return dp[n-1]
}
